src/Chainweb/Pact/Backend/RelationalCheckpointer.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE OverloadedStrings   #-}
+
+-- |
+-- Module: Chainweb.Pact.PactService
+-- Copyright: Copyright © 2018 Kadena LLC.
+-- License: See LICENSE file
+-- Maintainers: Emmanuel Denloye <emmanuel@kadena.io>
+-- Stability: experimental
+--
+-- Pact Checkpointer for Chainweb
+module Chainweb.Pact.Backend.RelationalCheckpointer where
+
+import Control.Concurrent.MVar
+import Control.Exception.Safe
+import Control.Lens
+import Control.Monad.Except
+import Control.Monad.Reader
+import Control.Monad.State
+
+import Data.Serialize hiding (get)
+
+import Pact.Types.Gas (GasEnv(..))
+import Pact.Types.Logger (Logger(..))
+
+-- pact
+import Pact.Types.SQLite
+import Pact.Types.Pretty
+import Pact.Types.Runtime
+-- import Pact.PersistPactDb (DbEnv(..))
+-- import Pact.Persist.SQLite
+
+-- chainweb
+import Chainweb.BlockHash
+import Chainweb.BlockHeader
+import Chainweb.Pact.Backend.ChainwebPactDb
+import Chainweb.Pact.Backend.Types
+
+initRelationalCheckpointer ::
+     Db -> Logger -> GasEnv -> IO CheckpointEnv
+initRelationalCheckpointer dbconn loggr gasEnv = do
+  let checkpointer =
+        CheckpointerNew
+          { restoreNew = innerRestore dbconn
+          , restoreInitialNew = innerRestoreInitial dbconn
+          , saveNew = innerSave dbconn
+          , saveInitialNew = innerSaveInitial dbconn
+          , discardNew = innerDiscard dbconn
+          }
+  return $
+    CheckpointEnv
+      { _cpeCheckpointer = undefined $ checkpointer
+      , _cpeLogger = loggr
+      , _cpeGasEnv = gasEnv
+      }
+
+type Db = MVar (CWDbEnv SQLiteEnv)
+
+innerRestore :: Db -> BlockHeight -> BlockHash -> IO (Either String (SQLiteEnv, TxId, Maybe ExecutionMode))
+innerRestore dbenv bh hsh = runCWDb dbenv $ runExceptT $ withExceptT ((mappend "restore :") . show) $ do
```

Comment:
ditch the ExceptT and just encase everything in a `tryAny`

Reply:
done